File: python/cutfemx/wrappers/extensions.cpp
// Copyright (c) 2026 ONERA
// Authors: Susanne Claus
// This file is part of CutFEMx
//
// SPDX-License-Identifier:    MIT

#include <array.h>

#include <complex>
#include <cstdint>
#include <nanobind/nanobind.h>
#include <nanobind/ndarray.h>
#include <nanobind/stl/complex.h>
#include <nanobind/stl/optional.h>
#include <nanobind/stl/shared_ptr.h>
#include <nanobind/stl/string.h>
#include <nanobind/stl/vector.h>
#include <string>

#include <cutfemx/cut/cut.h>
#include <cutfemx/extensions/cell_aggregation.h>
#include <cutfemx/extensions/extension_penalty.h>

#include <dolfinx/fem/FunctionSpace.h>
#include <dolfinx/la/MatrixCSR.h>

namespace nb = nanobind;

namespace
{
template <typename T>
void declare_extensions(nb::module_& m, const std::string& type)
{
  using CellAggregation = cutfemx::extensions::CellAggregation<T>;
  using ExtensionQuadrature = cutfemx::extensions::ExtensionQuadrature<T>;
  using CutData = cutfemx::CutData<T>;
  using FunctionSpace = dolfinx::fem::FunctionSpace<T>;
  using MatrixCSR = dolfinx::la::MatrixCSR<T>;

  std::string aggregation_name = "CellAggregation_" + type;
  nb::class_<CellAggregation>(m, aggregation_name.c_str(),
                              "CutFEMx cell aggregation data")
      .def_prop_ro(
          "active_cells",
          [](CellAggregation& self)
          {
            return nb::ndarray<const std::int32_t, nb::numpy>(
                self.active_cells.data(), {self.active_cells.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "cut_cells",
          [](CellAggregation& self)
          {
            return nb::ndarray<const std::int32_t, nb::numpy>(
                self.cut_cells.data(), {self.cut_cells.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "interior_cells",
          [](CellAggregation& self)
          {
            return nb::ndarray<const std::int32_t, nb::numpy>(
                self.interior_cells.data(), {self.interior_cells.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "well_posed_cells",
          [](CellAggregation& self)
          {
            return nb::ndarray<const std::int32_t, nb::numpy>(
                self.well_posed_cells.data(), {self.well_posed_cells.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "ill_posed_cells",
          [](CellAggregation& self)
          {
            return nb::ndarray<const std::int32_t, nb::numpy>(
                self.ill_posed_cells.data(), {self.ill_posed_cells.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "root_cell",
          [](CellAggregation& self)
          {
            return nb::ndarray<const std::int32_t, nb::numpy>(
                self.root_cell.data(), {self.root_cell.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "aggregate_id",
          [](CellAggregation& self)
          {
            return nb::ndarray<const std::int32_t, nb::numpy>(
                self.aggregate_id.data(), {self.aggregate_id.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "propagation_depth",
          [](CellAggregation& self)
          {
            return nb::ndarray<const std::int32_t, nb::numpy>(
                self.propagation_depth.data(), {self.propagation_depth.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "rootless_cells",
          [](CellAggregation& self)
          {
            return nb::ndarray<const std::int32_t, nb::numpy>(
                self.rootless_cells.data(), {self.rootless_cells.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "cut_volume_fraction",
          [](CellAggregation& self)
          {
            return nb::ndarray<const T, nb::numpy>(
                self.cut_volume_fraction.data(),
                {self.cut_volume_fraction.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal);

  std::string quadrature_name = "ExtensionQuadrature_" + type;
  nb::class_<ExtensionQuadrature>(m, quadrature_name.c_str(),
                                  "Full-cell bad/root extension quadrature")
      .def_prop_ro(
          "bad_cells",
          [](ExtensionQuadrature& self)
          {
            std::vector<std::int32_t> cells;
            cells.reserve(self.pairs.size());
            for (const auto& pair : self.pairs)
              cells.push_back(pair.bad_cell);
            return cells;
          })
      .def_prop_ro(
          "root_cells",
          [](ExtensionQuadrature& self)
          {
            std::vector<std::int32_t> cells;
            cells.reserve(self.pairs.size());
            for (const auto& pair : self.pairs)
              cells.push_back(pair.root_cell);
            return cells;
          })
      .def_prop_ro(
          "offsets",
          [](ExtensionQuadrature& self)
          {
            return nb::ndarray<const std::int32_t, nb::numpy>(
                self.offsets.data(), {self.offsets.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "bad_reference_points",
          [](ExtensionQuadrature& self)
          {
            return nb::ndarray<const T, nb::numpy>(
                self.bad_reference_points.data(),
                {self.bad_reference_points.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "root_reference_points",
          [](ExtensionQuadrature& self)
          {
            return nb::ndarray<const T, nb::numpy>(
                self.root_reference_points.data(),
                {self.root_reference_points.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "weights",
          [](ExtensionQuadrature& self)
          {
            return nb::ndarray<const T, nb::numpy>(
                self.weights.data(), {self.weights.size()},
                nb::cast(self, nb::rv_policy::reference));
          },
          nb::rv_policy::reference_internal)
      .def_ro("tdim", &ExtensionQuadrature::tdim);

  m.def(
      ("create_cell_aggregation_" + type).c_str(),
      [](const CutData& cut_data, const std::string& selector,
         T volume_fraction_threshold, const std::string& root_policy,
         int max_iterations, bool allow_rootless)
      {
        return cutfemx::extensions::create_cell_aggregation(
            cut_data, selector, volume_fraction_threshold,
            cutfemx::extensions::root_policy_from_string(root_policy),
            max_iterations, allow_rootless);
      },
      nb::arg("cut_data"), nb::arg("selector"),
      nb::arg("volume_fraction_threshold"),
      nb::arg("root_policy") = "interior_or_well_cut",
      nb::arg("max_iterations") = -1, nb::arg("allow_rootless") = false);

  m.def(
      ("extension_quadrature_" + type).c_str(),
      [](std::shared_ptr<const FunctionSpace> V, const CutData& cut_data,
         const CellAggregation& aggregation, int quadrature_degree)
      {
        if (!V)
          throw std::runtime_error("Received a null function space.");
        return cutfemx::extensions::extension_quadrature(
            *V, cut_data, aggregation, quadrature_degree);
      },
      nb::arg("V"), nb::arg("cut_data"), nb::arg("aggregation"),
      nb::arg("quadrature_degree"),
      "Build standard full-cell quadrature on bad cells, pulled back to roots.");

  m.def(
      ("create_extension_penalty_matrix_" + type).c_str(),
      [](std::shared_ptr<const FunctionSpace> V, const CutData& cut_data,
         const CellAggregation& aggregation)
      {
        if (!V)
          throw std::runtime_error("Received a null function space.");
        return cutfemx::extensions::create_extension_penalty_matrix<T, T>(
            *V, cut_data, aggregation);
      },
      nb::arg("V"), nb::arg("cut_data"), nb::arg("aggregation"),
      "Create a MatrixCSR with sparsity for extension-penalty bad/root pairs.");

  m.def(
      ("assemble_extension_penalty_scalar_" + type).c_str(),
      [](MatrixCSR& A, std::shared_ptr<const FunctionSpace> V,
         const CutData& cut_data, const CellAggregation& aggregation, T beta,
         int quadrature_degree)
      {
        if (!V)
          throw std::runtime_error("Received a null function space.");
        cutfemx::extensions::assemble_extension_penalty(
            A, *V, cut_data, aggregation, beta, quadrature_degree);
      },
      nb::arg("A"), nb::arg("V"), nb::arg("cut_data"),
      nb::arg("aggregation"), nb::arg("beta"),
      nb::arg("quadrature_degree"),
      "Assemble scalar-coefficient extension penalty into an existing MatrixCSR.");

  m.def(
      ("assemble_extension_penalty_cellwise_" + type).c_str(),
      [](MatrixCSR& A, std::shared_ptr<const FunctionSpace> V,
         const CutData& cut_data, const CellAggregation& aggregation,
         nb::ndarray<const T, nb::ndim<1>, nb::c_contig> beta_cell_values,
         int quadrature_degree)
      {
        if (!V)
          throw std::runtime_error("Received a null function space.");
        cutfemx::extensions::assemble_extension_penalty(
            A, *V, cut_data, aggregation,
            std::span<const T>(beta_cell_values.data(),
                               beta_cell_values.size()),
            quadrature_degree);
      },
      nb::arg("A"), nb::arg("V"), nb::arg("cut_data"),
      nb::arg("aggregation"), nb::arg("beta_cell_values"),
      nb::arg("quadrature_degree"),
      "Assemble cellwise-coefficient extension penalty into an existing MatrixCSR.");

  m.def(
      ("extension_penalty_matrix_scalar_" + type).c_str(),
      [](std::shared_ptr<const FunctionSpace> V, const CutData& cut_data,
         const CellAggregation& aggregation, T beta, int quadrature_degree)
      {
        if (!V)
          throw std::runtime_error("Received a null function space.");
        return cutfemx::extensions::extension_penalty_matrix(
            *V, cut_data, aggregation, beta, quadrature_degree);
      },
      nb::arg("V"), nb::arg("cut_data"), nb::arg("aggregation"),
      nb::arg("beta"), nb::arg("quadrature_degree"),
      "Create and assemble a scalar-coefficient extension penalty MatrixCSR.");

  m.def(
      ("extension_penalty_matrix_cellwise_" + type).c_str(),
      [](std::shared_ptr<const FunctionSpace> V, const CutData& cut_data,
         const CellAggregation& aggregation,
         nb::ndarray<const T, nb::ndim<1>, nb::c_contig> beta_cell_values,
         int quadrature_degree)
      {
        if (!V)
          throw std::runtime_error("Received a null function space.");
        return cutfemx::extensions::extension_penalty_matrix(
            *V, cut_data, aggregation,
            std::span<const T>(beta_cell_values.data(),
                               beta_cell_values.size()),
            quadrature_degree);
      },
      nb::arg("V"), nb::arg("cut_data"), nb::arg("aggregation"),
      nb::arg("beta_cell_values"),
      nb::arg("quadrature_degree"),
      "Create and assemble a cellwise-coefficient extension penalty MatrixCSR.");
}

template <typename Scalar, typename Geometry>
void declare_extension_matrices(nb::module_& m, const std::string& type)
{
  using CellAggregation = cutfemx::extensions::CellAggregation<Geometry>;
  using CutData = cutfemx::CutData<Geometry>;
  using FunctionSpace = dolfinx::fem::FunctionSpace<Geometry>;
  using MatrixCSR = dolfinx::la::MatrixCSR<Scalar>;

  m.def(
      ("create_extension_penalty_matrix_" + type).c_str(),
      [](std::shared_ptr<const FunctionSpace> V, const CutData& cut_data,
         const CellAggregation& aggregation)
      {
        if (!V)
          throw std::runtime_error("Received a null function space.");
        return cutfemx::extensions::create_extension_penalty_matrix<Scalar, Geometry>(
            *V, cut_data, aggregation);
      },
      nb::arg("V"), nb::arg("cut_data"), nb::arg("aggregation"),
      "Create a MatrixCSR with extension-penalty sparsity.");

  m.def(
      ("assemble_extension_penalty_scalar_" + type).c_str(),
      [](MatrixCSR& A, std::shared_ptr<const FunctionSpace> V,
         const CutData& cut_data, const CellAggregation& aggregation,
         Scalar beta, int quadrature_degree)
      {
        if (!V)
          throw std::runtime_error("Received a null function space.");
        cutfemx::extensions::assemble_extension_penalty(
            A, *V, cut_data, aggregation, beta, quadrature_degree);
      },
      nb::arg("A"), nb::arg("V"), nb::arg("cut_data"),
      nb::arg("aggregation"), nb::arg("beta"),
      nb::arg("quadrature_degree"),
      "Assemble scalar-coefficient extension penalty into MatrixCSR.");

  m.def(
      ("assemble_extension_penalty_cellwise_" + type).c_str(),
      [](MatrixCSR& A, std::shared_ptr<const FunctionSpace> V,
         const CutData& cut_data, const CellAggregation& aggregation,
         nb::ndarray<const Scalar, nb::ndim<1>, nb::c_contig> beta_cell_values,
         int quadrature_degree)
      {
        if (!V)
          throw std::runtime_error("Received a null function space.");
        cutfemx::extensions::assemble_extension_penalty(
            A, *V, cut_data, aggregation,
            std::span<const Scalar>(beta_cell_values.data(),
                                    beta_cell_values.size()),
            quadrature_degree);
      },
      nb::arg("A"), nb::arg("V"), nb::arg("cut_data"),
      nb::arg("aggregation"), nb::arg("beta_cell_values"),
      nb::arg("quadrature_degree"),
      "Assemble cellwise-coefficient extension penalty into MatrixCSR.");

  m.def(
      ("extension_penalty_matrix_scalar_" + type).c_str(),
      [](std::shared_ptr<const FunctionSpace> V, const CutData& cut_data,
         const CellAggregation& aggregation, Scalar beta, int quadrature_degree)
      {
        if (!V)
          throw std::runtime_error("Received a null function space.");
        return cutfemx::extensions::extension_penalty_matrix(
            *V, cut_data, aggregation, beta, quadrature_degree);
      },
      nb::arg("V"), nb::arg("cut_data"), nb::arg("aggregation"),
      nb::arg("beta"), nb::arg("quadrature_degree"),
      "Create and assemble a scalar-coefficient MatrixCSR.");

  m.def(
      ("extension_penalty_matrix_cellwise_" + type).c_str(),
      [](std::shared_ptr<const FunctionSpace> V, const CutData& cut_data,
         const CellAggregation& aggregation,
         nb::ndarray<const Scalar, nb::ndim<1>, nb::c_contig> beta_cell_values,
         int quadrature_degree)
      {
        if (!V)
          throw std::runtime_error("Received a null function space.");
        return cutfemx::extensions::extension_penalty_matrix(
            *V, cut_data, aggregation,
            std::span<const Scalar>(beta_cell_values.data(),
                                    beta_cell_values.size()),
            quadrature_degree);
      },
      nb::arg("V"), nb::arg("cut_data"), nb::arg("aggregation"),
      nb::arg("beta_cell_values"),
      nb::arg("quadrature_degree"),
      "Create and assemble a cellwise-coefficient MatrixCSR.");
}
} // namespace

namespace cutfemx_wrappers
{
void extensions(nb::module_& m)
{
  declare_extensions<float>(m, "float32");
  declare_extensions<double>(m, "float64");
  declare_extension_matrices<float, double>(m, "float32_float64");
  declare_extension_matrices<double, float>(m, "float64_float32");
  declare_extension_matrices<std::complex<float>, float>(
      m, "complex64_float32");
  declare_extension_matrices<std::complex<float>, double>(
      m, "complex64_float64");
  declare_extension_matrices<std::complex<double>, float>(
      m, "complex128_float32");
  declare_extension_matrices<std::complex<double>, double>(
      m, "complex128_float64");
}
} // namespace cutfemx_wrappers

File: python/cutfemx/extensions.py
# Copyright (c) 2026 ONERA
# Authors: Susanne Claus
# This file is part of CutFEMx
#
# SPDX-License-Identifier:    MIT

from __future__ import annotations

from dataclasses import dataclass
import numbers
from typing import Any

import numpy as np
import numpy.typing as npt

from cutfemx import cutfemx_cpp as _cpp
from cutfemx.cut import CutData


@dataclass
class CellAggregation:
    """Python handle for CutFEMx cell aggregation data."""

    _cpp_object: Any

    @property
    def active_cells(self):
        return self._cpp_object.active_cells

    @property
    def cut_cells(self):
        return self._cpp_object.cut_cells

    @property
    def interior_cells(self):
        return self._cpp_object.interior_cells

    @property
    def well_posed_cells(self):
        return self._cpp_object.well_posed_cells

    @property
    def ill_posed_cells(self):
        return self._cpp_object.ill_posed_cells

    @property
    def root_cell(self):
        return self._cpp_object.root_cell

    @property
    def aggregate_id(self):
        return self._cpp_object.aggregate_id

    @property
    def propagation_depth(self):
        return self._cpp_object.propagation_depth

    @property
    def rootless_cells(self):
        return self._cpp_object.rootless_cells

    @property
    def cut_volume_fraction(self):
        return self._cpp_object.cut_volume_fraction


@dataclass
class ExtensionQuadrature:
    """Full-cell bad/root quadrature for aggregation extension terms."""

    _cpp_object: Any

    @property
    def bad_cells(self):
        return self._cpp_object.bad_cells

    @property
    def root_cells(self):
        return self._cpp_object.root_cells

    @property
    def offsets(self):
        return self._cpp_object.offsets

    @property
    def bad_reference_points(self):
        return self._cpp_object.bad_reference_points

    @property
    def root_reference_points(self):
        return self._cpp_object.root_reference_points

    @property
    def weights(self):
        return self._cpp_object.weights

    @property
    def tdim(self):
        return self._cpp_object.tdim


@dataclass
class ExtensionPenaltyTerm:
    """One PDE-dependent aggregation extension penalty contribution."""

    V: Any
    beta: Any
    quadrature_degree: int
    product: str = "L2"
    cut_data: Any | None = None
    aggregation: Any | None = None

    def __post_init__(self):
        if self.product != "L2":
            raise NotImplementedError("Only L2 extension penalty terms are implemented in v1")

    def with_domain(
        self, cut_data: CutData, aggregation: CellAggregation
    ) -> "ExtensionPenaltyTerm":
        """Return a copy bound to the cut data and aggregation used for assembly."""
        return ExtensionPenaltyTerm(
            self.V,
            self.beta,
            self.quadrature_degree,
            self.product,
            cut_data,
            aggregation,
        )


def create_cell_aggregation(
    cut_data: CutData,
    selector: str,
    volume_fraction_threshold: float,
    *,
    root_policy: str = "interior_or_well_cut",
    max_iterations: int = -1,
    allow_rootless: bool = False,
) -> CellAggregation:
    """Create serial v1 cell aggregation data from CutFEMx cut data."""
    if not isinstance(cut_data, CutData):
        raise TypeError("create_cell_aggregation expects a cutfemx.CutData object")
    if not hasattr(_cpp, "extensions"):
        raise RuntimeError("CutFEMx was built without extensions support")

    dtype = cut_data.mesh.geometry.x.dtype
    if dtype.name == "float32":
        cpp_object = _cpp.extensions.create_cell_aggregation_float32(
            cut_data._cpp_object,
            selector,
            volume_fraction_threshold,
            root_policy,
            max_iterations,
            allow_rootless,
        )
    else:
        cpp_object = _cpp.extensions.create_cell_aggregation_float64(
            cut_data._cpp_object,
            selector,
            volume_fraction_threshold,
            root_policy,
            max_iterations,
            allow_rootless,
        )
    return CellAggregation(cpp_object)


def _geometry_type_name(V) -> str:
    dtype = V.mesh.geometry.x.dtype
    return "float32" if dtype.name == "float32" else "float64"


def _scalar_type_name(dtype: Any) -> str:
    dtype = np.dtype(dtype)
    if dtype == np.dtype(np.float32):
        return "float32"
    if dtype == np.dtype(np.float64):
        return "float64"
    if dtype == np.dtype(np.complex64):
        return "complex64"
    if dtype == np.dtype(np.complex128):
        return "complex128"
    raise NotImplementedError(f"Unsupported extension penalty scalar dtype {dtype}.")


def _matrix_scalar_dtype(A: Any) -> np.dtype:
    type_name = type(A._cpp_object).__name__
    if "complex64" in type_name:
        return np.dtype(np.complex64)
    if "complex128" in type_name:
        return np.dtype(np.complex128)
    if "float32" in type_name:
        return np.dtype(np.float32)
    if "float64" in type_name:
        return np.dtype(np.float64)
    raise TypeError(f"Unsupported MatrixCSR scalar type {type_name!r}.")


def _extension_suffix(scalar_type: str, geometry_type: str) -> str:
    return (
        geometry_type
        if scalar_type == geometry_type
        else f"{scalar_type}_{geometry_type}"
    )


def _extension_function(base: str, scalar_dtype: Any, V: Any) -> Any:
    scalar_type = _scalar_type_name(scalar_dtype)
    geometry_type = _geometry_type_name(V)
    attr = f"{base}_{_extension_suffix(scalar_type, geometry_type)}"
    if not hasattr(_cpp.extensions, attr):
        raise RuntimeError(
            "CutFEMx was built without extension penalty support for "
            f"{scalar_type} matrices on {geometry_type} geometry."
        )
    return getattr(_cpp.extensions, attr)


def _beta_scalar_dtype(beta: Any, mesh) -> np.dtype:
    if isinstance(beta, np.ndarray):
        return np.dtype(beta.dtype)
    if isinstance(beta, numbers.Number):
        return np.dtype(np.asarray(beta).dtype)
    if hasattr(beta, "x") and hasattr(beta.x, "array"):
        return np.dtype(beta.x.array.dtype)
    return np.dtype(mesh.geometry.x.dtype)


def _cellwise_values(beta: Any, mesh, dtype: np.dtype) -> npt.NDArray:
    if isinstance(beta, np.ndarray):
        values = np.asarray(beta, dtype=dtype)
        if values.ndim != 1:
            raise ValueError("Cellwise extension penalty beta array must be one-dimensional")
        return np.ascontiguousarray(values)

    if not (hasattr(beta, "function_space") and hasattr(beta, "x")):
        raise TypeError(
            "Extension penalty beta must be a scalar, numpy cell array, or DG0 Function"
        )

    Vb = beta.function_space
    if Vb.mesh is not mesh:
        raise ValueError("Extension penalty beta Function must live on the same mesh")

    tdim = mesh.topology.dim
    num_cells = mesh.topology.index_map(tdim).size_local
    values = np.empty(num_cells, dtype=dtype)
    for cell in range(num_cells):
        dofs = Vb.dofmap.cell_dofs(cell)
        if len(dofs) != 1:
            raise ValueError(
                "Extension penalty beta Function must have exactly one dof per cell "
                "(DG0 is supported in v1)"
            )
        values[cell] = beta.x.array[dofs[0]]
    return np.ascontiguousarray(values)


def create_extension_penalty_matrix(
    V,
    cut_data: CutData,
    aggregation: CellAggregation,
    dtype: Any | None = None,
):
    """Create a MatrixCSR with sparsity for bad/root extension penalty pairs."""
    if isinstance(V, ExtensionPenaltyTerm):
        V = V.V
    if not isinstance(cut_data, CutData):
        raise TypeError("create_extension_penalty_matrix expects a cutfemx.CutData object")
    if not isinstance(aggregation, CellAggregation):
        raise TypeError("create_extension_penalty_matrix expects a CellAggregation object")
    scalar_dtype = V.mesh.geometry.x.dtype if dtype is None else np.dtype(dtype)
    return _extension_function("create_extension_penalty_matrix", scalar_dtype, V)(
        V._cpp_object, cut_data._cpp_object, aggregation._cpp_object
    )


def extension_quadrature(
    V,
    cut_data: CutData,
    aggregation: CellAggregation,
    quadrature_degree: int,
) -> ExtensionQuadrature:
    """Build standard full-cell quadrature on bad cells and pull it to roots."""
    if isinstance(V, ExtensionPenaltyTerm):
        quadrature_degree = V.quadrature_degree
        V = V.V
    if not isinstance(cut_data, CutData):
        raise TypeError("extension_quadrature expects a cutfemx.CutData object")
    if not isinstance(aggregation, CellAggregation):
        raise TypeError("extension_quadrature expects a CellAggregation object")
    type_name = _geometry_type_name(V)
    cpp_object = getattr(_cpp.extensions, f"extension_quadrature_{type_name}")(
        V._cpp_object,
        cut_data._cpp_object,
        aggregation._cpp_object,
        quadrature_degree,
    )
    return ExtensionQuadrature(cpp_object)


def assemble_extension_penalty(
    A,
    V,
    cut_data: CutData,
    aggregation: CellAggregation,
    beta: Any | None = None,
    quadrature_degree: int | None = None,
):
    """Assemble an L2 full-bad-cell extension penalty into ``A``.

    ``beta`` may be a scalar, a one-dimensional cellwise numpy array, or a
    scalar DG0 ``dolfinx.fem.Function``. Function coefficients are evaluated on
    the bad cell. The quadrature domain is the full background bad cell, while
    the same physical quadrature points are also evaluated in root coordinates.
    """
    if not isinstance(cut_data, CutData):
        raise TypeError("assemble_extension_penalty expects a cutfemx.CutData object")
    if not isinstance(aggregation, CellAggregation):
        raise TypeError("assemble_extension_penalty expects a CellAggregation object")
    if isinstance(V, ExtensionPenaltyTerm):
        if beta is not None or quadrature_degree is not None:
            raise ValueError("Do not pass beta/quadrature_degree when using ExtensionPenaltyTerm")
        term = V
        V = term.V
        beta = term.beta
        quadrature_degree = term.quadrature_degree
    if beta is None or quadrature_degree is None:
        raise TypeError("assemble_extension_penalty requires beta and quadrature_degree")

    scalar_dtype = _matrix_scalar_dtype(A)
    if isinstance(beta, numbers.Number):
        _extension_function("assemble_extension_penalty_scalar", scalar_dtype, V)(
            A._cpp_object,
            V._cpp_object,
            cut_data._cpp_object,
            aggregation._cpp_object,
            scalar_dtype.type(beta),
            quadrature_degree,
        )
    else:
        values = _cellwise_values(beta, V.mesh, scalar_dtype)
        _extension_function("assemble_extension_penalty_cellwise", scalar_dtype, V)(
            A._cpp_object,
            V._cpp_object,
            cut_data._cpp_object,
            aggregation._cpp_object,
            values,
            quadrature_degree,
        )
    return A


def extension_penalty_matrix(
    V,
    cut_data: CutData,
    aggregation: CellAggregation,
    beta: Any | None = None,
    quadrature_degree: int | None = None,
):
    """Create and assemble one L2 full-bad-cell extension penalty MatrixCSR."""
    if not isinstance(cut_data, CutData):
        raise TypeError("extension_penalty_matrix expects a cutfemx.CutData object")
    if not isinstance(aggregation, CellAggregation):
        raise TypeError("extension_penalty_matrix expects a CellAggregation object")
    if isinstance(V, ExtensionPenaltyTerm):
        if beta is not None or quadrature_degree is not None:
            raise ValueError("Do not pass beta/quadrature_degree when using ExtensionPenaltyTerm")
        term = V
        V = term.V
        beta = term.beta
        quadrature_degree = term.quadrature_degree
    if beta is None or quadrature_degree is None:
        raise TypeError("extension_penalty_matrix requires beta and quadrature_degree")

    scalar_dtype = _beta_scalar_dtype(beta, V.mesh)
    if isinstance(beta, numbers.Number):
        return _extension_function("extension_penalty_matrix_scalar", scalar_dtype, V)(
            V._cpp_object,
            cut_data._cpp_object,
            aggregation._cpp_object,
            scalar_dtype.type(beta),
            quadrature_degree,
        )

    values = _cellwise_values(beta, V.mesh, scalar_dtype)
    return _extension_function("extension_penalty_matrix_cellwise", scalar_dtype, V)(
        V._cpp_object,
        cut_data._cpp_object,
        aggregation._cpp_object,
        values,
        quadrature_degree,
    )
